### DISCUSSION TITLE
dotnet core 3.1 (Docker)

### DISCUSSION THREAD

**Discussion by WoodsterDK** [dbosoft/YaNco#163] [Q&A] [answered]
Hi there,

I seem to have trouble getting it to run from a dotnet core 3.1 docker image.
If I run the app it works, but as soon as it is being dockerized, it doesn't work.

Missing dependencies or any ideas?

**Comment by fw2568** [comment, marked as answer]
Hi,

without any details it is difficult to gues what kind of problems you had, but in general it is working. But as YaNco uses the native SAP RFC SDK you have to consider its requirements for a docker container. 
Beside using the linux native libraries within docker, you have to take care of the LD_LIBRARY_PATH variable, that has to point to the sdk binaries. 
Here you can find a working sample application that is using also docker: https://github.com/dbosoft/SAPHub

dockerfile: https://github.com/dbosoft/SAPHub/blob/main/src/SAPHub.SAPConnector/Dockerfile

Please notice the ENV statement that sets the LD_LIBRARY_PATH to the container's /app directory:

https://github.com/dbosoft/SAPHub/blob/8e9d6c1d47e6a053c5ba579d4d4df41e50116369/src/SAPHub.SAPConnector/Dockerfile#L21-L24

Hope that helps. If you still facing problems please provide a example with source. 

Best Regards,
Frank

**Reply by WoodsterDK** [reply]
Hi Frank,
Thanks for taking the time to answer.

I did set the library path variable, but sounds like the issue could be caused by only referencing the x64 Win RFC SDK library.
I have only had that available, since the docker thing is a new feature I'm looking into. Will try to request the linux SDK files.

When doing the debug from a core 3.1 project, it seems there was a build reference to a version 5.0.0.0 lib from the yanco lib, and it would only work if I changed my project to dotnet 5. Can't remember which one - System.*.*Flow perhaps, but will get back later once I have had the chance to check.

Once again thanks....

**Reply by WoodsterDK** [reply]
Just did a check - and if running pure dotnet core 3.1, I get this exception:

Exception thrown: 'System.IO.FileNotFoundException' in Dbosoft.Functional.dll
Could not load file or assembly 'System.Threading.Tasks.Dataflow, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.

But if I add the net5.0 it works:

<TargetFrameworks>net5.0;netcoreapp3.1</TargetFrameworks>


@fw2568 One additional question - does the YaNco together with SAP RFC SDK support encryption from docker - SNC/sapcrypto connection?

**Reply by fw2568** [reply]
This is the second time during the last days someone mentioned a dependency issue with System.Threading.Tasks.Dataflow. I will check if it is necessary to lower the dep. requirements. 

Never tried using sapcrypto within docker, but as all crypto is implemented within sap components I would expect it to work at least for X.509 certificates.

**Reply by fw2568** [reply]
@WoodsterDK I tried to reproduce the issue with System.Threading.Tasks.Dataflow with a netcore 3.1 console app. However - both for Windows and Docker on Linux the version 5.0 of System.Threading.Tasks.Dataflow is selected as expected. 

I also checked requirements of System.Threading.Tasks.Dataflow. version 5.0 is marked as compatible with all frameworks (see https://www.nuget.org/packages/System.Threading.Tasks.Dataflow/5.0.0). So there should be no issue expect if another library is included that forces to include a lower version of DataFlow. But in that case you should get a compatibility warning on package restore.

It would help a lot if you can share a sample project with all dependencies included as in the app where you had the issue (no code needed). 

Thanks,
Frank

**Reply by WoodsterDK** [reply]
@fw2568 thanks for looking into it.

I'm out of office the next couple of days, but for sure will look into it, and get back.

Can you do a stand alone core 3.1 example app to test with as well...?

Again, Thanks...

**Reply by WoodsterDK** [reply]
@fw2568 Hi Frank,
I just received the Linux files and it seems to be working with a dotnet 5 app -> docker image.
When running the docker image the connection is successful.

When doing it as a dotnet core 3.1 app - I still get this error:
"System.IO.FileNotFoundException: 'Could not load file or assembly 'System.Threading.Tasks.Dataflow, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.'"

And that is when debugging from Visual Studio.

Could you try to do a simple dotnet core 3.1 I can try to use as templete - to see if I can figure out what is wrong?

**Reply by fw2568** [reply]
Could you please share the code?
Here a sample app where it is working with netcore 3.1: https://github.com/fw2568/YancoAzureFunctionTest/tree/master/YaNCo_netcore31

**Reply by WoodsterDK** [reply]
@fw2568 Just tried your test app, and that seems to be working when debugging dotnet core from Visual Studio.
I must have messed something up in the project - will try to start from scratch....

Thanks for you help so far....

**Reply by fw2568** [reply]
For some reason final solution is not visible in discussion, adding it here for documentation: 

reply from **WoodsterDK**: 

OK - I didn't use the nuget dependencies, and just had the System.Threading.Tasks.Dataflow.dll copied to the output folder - BUT it seems only to work for dotnet core 3.1, if it is added as a reference in the project, and not just present in the folder.
So it seems to work - WUHOO...
